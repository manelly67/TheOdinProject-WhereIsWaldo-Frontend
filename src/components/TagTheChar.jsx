import { useState, useEffect } from "react";
import { fromNormalizeToScreen } from "./coord";

const TagTheChar = (props) => {
  const { imgCharacters, W, H, coords } = props;
  const [imgCharacToTag, setImgCharacToTag] = useState(null);

  useEffect(() => {
    let temp = [];
    imgCharacters.forEach((e) => {
      if (e.found === true) {
        const { x, y } = fromNormalizeToScreen(
          Number(e.x),
          Number(e.y),
          W,
          H,
          coords
        );
        temp.push({ id: e.id, name: e.name, found: e.found, x: x, y: y });
      } else {
        temp.push({ id: e.id, name: e.name, found: e.found, x: e.x, y: e.y });
      }
    });
    console.log(temp);
    setImgCharacToTag(temp);
  }, [imgCharacters, W, H, coords]);

  const allTags =
    imgCharacToTag === null
      ? null
      : imgCharacToTag.map((e) => (
          <div
            key={e.id}
            style={{
             /*  backgroundColor: "black",
              color: "white",
              fontSize: "0.6rem",
              border: "1px solid white", */
              zIndex: "1",
            }}
          >
            {e.found === false ? null : (
              <p
                key={crypto.randomUUID()}
                role="tag"
                style={{
                  position: "absolute",
                  top: `${e.y + 5 }px`,
                  left: `${e.x  }px`,
                  backgroundColor: "black",
                  border: "1px solid white",
                  color: "yellow",
                  fontSize: "0.6rem",
                  fontWeight: 900,
                }}
              >
                {e.name}
              </p>
            )}
          </div>
        ));

  return (
  <>
  {imgCharacToTag ? <div>{allTags}</div> : null}
  </>
  );
};

export default TagTheChar;
