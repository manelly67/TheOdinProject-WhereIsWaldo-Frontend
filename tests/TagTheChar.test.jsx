import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";

import TagTheChar from "../src/components/TagTheChar.jsx";

describe("simple use of mock fuction", () => {
  it("mock function", () => {
    const fn = vi.fn();
    fn("hello world");

    expect(fn.mock.calls[0]).toEqual(["hello world"]);

    const market = {
      getApples: () => 100,
    };

    const getApplesSpy = vi.spyOn(market, "getApples");
    market.getApples();

    expect(getApplesSpy.mock.calls.length).toEqual(1);
    expect(market.getApples()).toBe(100);
  });
});

describe("display the tag of characters", () => {
    const W = 900;
    const H = 500;
    const coords = {x:36,y:138};
   
  it("tags when no characters have been founded", () => {
    const imgCharacters = [
      { id: "char-1", name: "Wally", found: false, x: 0, y: 0 },
      { id: "char-2", name: "The Wizard", found: false, x: 0, y: 0 },
      {
        id: "char-3",
        name: "Courage the Cowardly Dog",
        found: false,
        x: 0,
        y: 0,
      },
    ];

    render(<TagTheChar imgCharacters={imgCharacters} coords={coords} W={W} H={H} />);
    expect(screen.queryAllByRole("tag").length).toBe(0);
  });

  it("tags when 1 character have been founded", () => {
   
    const imgCharacters = [
      { id: "char-1", name: "Wally", found: false, x: 0, y: 0 },
      { id: "char-2", name: "The Wizard", found: false, x: 0, y: 0 },
      {
        id: "char-3",
        name: "Courage the Cowardly Dog",
        found: true,
        x: -0.1586304817,
        y: -0.6217662801,
      },
    ];

    const {container} = render(<TagTheChar imgCharacters={imgCharacters} coords={coords} W={W} H={H} />);
    expect(container).toMatchSnapshot();
    expect(screen.queryAllByRole("tag").length).toBe(1);
    expect(screen.getByText(/Courage the Cowardly Dog/i)).toBeInTheDocument();
    expect(screen.queryByText(/The Wizard/i)).not.toBeInTheDocument();

  });

});
