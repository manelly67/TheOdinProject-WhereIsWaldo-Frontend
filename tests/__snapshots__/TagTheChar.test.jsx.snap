// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`display the tag of characters > tags when 1 character have been founded 1`] = `
<div>
  <div>
    <div
      style="z-index: 1;"
    />
    <div
      style="z-index: 1;"
    />
    <div
      style="z-index: 1;"
    >
      <p
        role="tag"
        style="position: absolute; top: 237.56px; left: 414.62px; background-color: black; border: 1px solid white; color: yellow; font-size: 0.6rem; font-weight: 900;"
      >
        Courage the Cowardly Dog
      </p>
    </div>
  </div>
</div>
`;
